ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/Worker.java

```diff
@@ -118,20 +120,33 @@ public void run() {
           singleRun.cancel(true);
           executor.shutdownNow();
           executor = getTimeoutHandlingExecutor();
+          err = true;
         } catch (ExecutionException e) {
           LOG.info("Exception during executing compaction", e);
+          err = true;
         } catch (InterruptedException ie) {
           // do not ignore interruption requests
           return;
+        } catch (Throwable t) {
+          err = true;
         }
 
         doPostLoopActions(System.currentTimeMillis() - startedAt);
 
         // If we didn't try to launch a job it either means there was no work to do or we got
-        // here as the result of a communication failure with the DB.  Either way we want to wait
+        // here as the result of an error like communication failure with the DB, schema failures etc.  Either way we want to wait
         // a bit before, otherwise we can start over the loop immediately.
-        if (!launchedJob && !stop.get()) {
+        if ((!launchedJob || err) && !stop.get()) {
           Thread.sleep(SLEEP_TIME);
+          //Backoff mechanism
+          //Increase sleep time if error persist
+          if(err && SLEEP_TIME < SLEEP_TIME_MAX) {
+            SLEEP_TIME *= 2;
```

Comment:
maybe
````
nextSleep = (err) ? nextSleep*2 : SLEEP_TIME;
if (nextSleep > MAX_SLEEP) nextSleep = MAX_SLEEP;
````

Reply:
I thought about using ternary, then I realised the current implementation is simple and only requires one condition check. Do you have any thing specific in mind with the new nextSleep variable ?

Reply:
SLEEP_TIME shouldn't be > MAX_SLEEP, but with current logic it could be. no preference to ternary, up to you

Reply:
Made the changes, thanks for the catch @deniskuzZ

Reply:
i think the below code would be easier to comprehend, WDYT?

````
long nextSleep = SLEEP_TIME;
..........
if ((!launchedJob || err) && !stop.get()) {
     Thread.sleep(nextSleep);
}
nextSleep = (err) ? nextSleep*2 : SLEEP_TIME; 
if (nextSleep > MAX_SLEEP) nextSleep = MAX_SLEEP;
````